forui/lib/src/widgets/pagination/pagination_controller.dart

```diff
@@ -0,0 +1,192 @@
+import 'dart:math';
+
+import 'package:flutter/widgets.dart';
+
+import 'package:meta/meta.dart';
+
+/// A controller that controls which page is selected.
+class FPaginationController extends PageController {
+  /// The number of sibling pages displayed beside the current page number. Defaults to 1.
+  ///
+  /// This value determines how many pages are shown on either side of the
+  /// currently selected page in the pagination.
+  ///
+  /// For example, if `siblings` is 2 and the current page is 5,
+  /// the displayed pages would be 3, 4, 5, 6, and 7.
+  ///
+  /// # Contract:
+  /// * Throws [AssertionError] if [siblings] < 0.
+  final int siblings;
+
+  /// Whether to show the first and last pages in the pagination. Defaults to `true`.
+  ///
+  /// If `true`, the pagination will always display the first and last page, regardless of the current page.
+  /// This can be useful for allowing users to quickly navigate to the beginning or end of the paginated content.
+  final bool showEdges;
+
+  /// The total number of pages.
+  ///
+  /// # Contract:
+  /// * Throws [AssertionError] if [length] < 0.
+  final int _length;
+  int _value;
+
+  /// Creates a [FPaginationController].
+  ///
+  /// # Contract:
+  /// * Throws [AssertionError] if 1 <= [page] and [page] <= length.
+  FPaginationController({
+    int? length,
+    int page = 0,
+    this.showEdges = true,
+    this.siblings = 1,
+    super.onAttach,
+    super.onDetach,
+    super.keepPage,
+    super.viewportFraction,
+  }) : assert(length == null || 0 < length, 'The total length of pages should be more than 0, but is $length.'),
+       assert(0 <= siblings, 'The siblingLength should be non-negative, but is $siblings'),
+       assert(
+         0 <= page && page < (length ?? 0),
+         'The initial page must be greater than or equal to 0 and less than or equal to length.',
+       ),
+       _value = page,
+       _length = length ?? 0,
+       super(initialPage: page) {
+    addListener(_handleScrollChanges);
+  }
+
+  void _handleScrollChanges() {
+    if (hasClients) {
+      if (page != null && page!.round() != _value) {
```

Comment:
this can be greatly simplified using the if(... case final page? when ...) syntax.